hub/src/server/utils.ts

```diff
@@ -63,3 +62,36 @@ export function timeout(milliseconds: number): Promise<void> {
     }, milliseconds)
   })
 }
+
+
+export class AsyncMutexScope {
+
+  private readonly _opened = new Set<string>()
+
+  public get openedCount() {
+    return this._opened.size
+  }
+
+  /**
+   * If no mutex of the given `id` is already taken, then a mutex is created and the 
+   * given promise is invoked. The mutex is released once the promise resolves -- either by 
+   * success or error. 
+   * @param id A unique mutex name used in a Map.
+   * @param spawnOwner A function that creates a Promise if the mutex is acquired. 
+   * @returns `true` if the mutex was acquired, otherwise returns `false`
+   */
+  public tryAcquire(id: string, spawnOwner: () => Promise<void>): boolean {
+    if (this._opened.has(id)) {
```

Comment:
We're _absolutely certain_ that this check will be atomic?

Reply:
Addressed in below comment. I just went through the lock path again and everything is normal synchronous js, and nothing else resulting in event loop ordering issues.